pkg/controller/plan/adapter/openstack/builder.go

```diff
@@ -61,17 +210,203 @@ func (r *Builder) VirtualMachine(vmRef ref.Ref, object *cnv.VirtualMachineSpec,
+		}
+		object.Template.Spec.Domain.Devices.Inputs = []cnv.Input{tablet}
+	}
+}
+
+func (r *Builder) mapCPUMem(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	// Set MachineType
+	machineType := DefaultProperties[MachineType]
+	if imageMachineType, ok := vm.Image.Properties[MachineType]; ok {
+		machineType = imageMachineType.(string)
+	} else {
+		for _, volume := range vm.Volumes {
+			if volume.Bootable == "true" {
+				if volumeMachineType, ok := volume.VolumeImageMetadata[MachineType]; ok {
+					machineType = volumeMachineType
+				}
+			}
+		}
+	}
+	object.Template.Spec.Domain.Machine = &cnv.Machine{Type: machineType}
+
+	// Set CPU Policy
+	cpuPolicy := DefaultProperties[CPUPolicy]
+	if imageCPUPolicy, ok := vm.Image.Properties[CPUPolicy]; ok {
+		cpuPolicy = imageCPUPolicy.(string)
+	}
+	if cpuPolicy == CPUPolicyDedicated {
+		object.Template.Spec.Domain.CPU.DedicatedCPUPlacement = true
+	}
+
+	// Set CPU Thread Policy
+	cpuThreadPolicy := DefaultProperties[CPUThreadPolicy]
+	if imageCPUThreadPolicy, ok := vm.Image.Properties[CPUThreadPolicy]; ok {
+		cpuThreadPolicy = imageCPUThreadPolicy.(string)
+	}
+	if cpuThreadPolicy == CPUThreadPolicyIsolate {
+		object.Template.Spec.Domain.CPU.IsolateEmulatorThread = true
+	}
+
+	// Set CPU Sockets/Cores/Threads and Memory requests
+	var cpuSockets, cpuCores, cpuThreads int
+
+	cpuSockets = vm.Flavor.VCPUs
+	cpuCores = 1
+	cpuThreads = 1
+
+	if flavorCPUSockets, ok := vm.Flavor.ExtraSpecs[FlavorCPUSockets]; ok {
+		if sockets, err := strconv.Atoi(flavorCPUSockets); err == nil {
+			cpuSockets = sockets
+		}
+	} else if imageCPUSockets, ok := vm.Image.Properties[CPUSockets]; ok {
+		if sockets, ok := imageCPUSockets.(int); ok {
+			cpuSockets = sockets
+		}
+	}
+	if flavorCPUCores, ok := vm.Flavor.ExtraSpecs[FlavorCPUCores]; ok {
+		if cores, err := strconv.Atoi(flavorCPUCores); err == nil {
+			cpuCores = cores
+		}
+	} else if imageCPUCores, ok := vm.Image.Properties[CPUCores]; ok {
+		if cores, ok := imageCPUCores.(int); ok {
+			cpuCores = cores
+		}
+	}
+	if flavorCPUThreads, ok := vm.Flavor.ExtraSpecs[FlavorCPUThreads]; ok {
+		if threads, err := strconv.Atoi(flavorCPUThreads); err == nil {
+			cpuThreads = threads
+		}
+	} else if imageCPUThreads, ok := vm.Image.Properties[CPUThreads]; ok {
+		if threads, ok := imageCPUThreads.(int); ok {
+			cpuThreads = threads
+		}
+	}
+
+	resourceRequests := map[core.ResourceName]resource.Quantity{}
+
+	if cpuCores != 1 || cpuThreads != 1 {
+		object.Template.Spec.Domain.CPU.Sockets = uint32(cpuSockets)
+		object.Template.Spec.Domain.CPU.Cores = uint32(cpuCores)
+		object.Template.Spec.Domain.CPU.Threads = uint32(cpuThreads)
+	} else {
+		cpuRequest := resource.NewQuantity(int64(cpuSockets*cpuCores*cpuThreads), resource.BinarySI)
+		resourceRequests[core.ResourceCPU] = *cpuRequest
+	}
+
+	memoryRequest := resource.NewQuantity(int64(vm.Flavor.RAM*int(resource.Mega)), resource.BinarySI)
+	resourceRequests[core.ResourceMemory] = *memoryRequest
```

Comment:
note the other part of the failure in CI:
spec.template.spec.domain.resources.requests.memory '3Ki': must be greater than or equal to 1M.

Reply:
I think I misunderstood the meaning of `resource.Mega`. This will be fixed in the next push

Reply:
Fixed